### PR TITLE
Make `IcuRatio::from_str` accept scientific notation numbers.

### PR DESCRIPTION
**Pull request by younies**
Make the IcuRatio::from_str accept any type of numbers from integer to scientific and rational numbers.

### CODE REVIEW

File: components/experimental/src/units/ratio.rs [hunk 1]
```diff
@@ -21,15 +21,28 @@ use super::provider::{Base, SiPrefix};
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub struct IcuRatio(Ratio<BigInt>);
 
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
 pub enum IcuRatioError {
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:25]: issue: we decided to not have catch-all error types going forward. Please make a custom error type for each function that can fail, which only contains error variants that are applicable to this function.

File: components/experimental/src/units/ratio.rs [hunk 2]
```diff
@@ -195,24 +208,184 @@ impl From<u32> for IcuRatio {
...
+        }
+
+        /// Parses a decimal number from a string input.
+        /// Accepts input in various decimal formats, including:
+        /// - Whole numbers: "1"
+        /// - Decimal numbers: "1.5"
+        /// An empty string input is interpreted as "0".
+        /// Note: Fractional inputs are not supported in this context.
+        fn parse_decimal(decimal: &str) -> Result<IcuRatio, IcuRatioError> {
+            let dot_count = decimal.matches('.').count();
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:270]: issue: this scans the string multiple times. Instead, do. `.split('.')` and call `.next()` a few times

File: components/experimental/src/units/ratio.rs [hunk 3]
```diff
@@ -195,24 +208,184 @@ impl From<u32> for IcuRatio {
...
+            match dot_count {
+                0 => BigRational::from_str(decimal)
+                    .map(IcuRatio)
+                    .map_err(|_| IcuRatioError::InvalidRatioString),
+                1 => {
+                    let decimal_part_length = decimal.split('.').last().unwrap().len();
+                    // remove "." from the string
+                    let number_str = decimal.replace(".", "");
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:278]: issue: `replace` allocates a new string. also below

File: components/experimental/src/units/ratio.rs [hunk 4]
```diff
@@ -195,24 +208,184 @@ impl From<u32> for IcuRatio {
...
         if parts.next().is_some() {
-            return Err(IcuRatioError::InvalidRatioString);
+            return Err(IcuRatioError::MultipleScientificNotations);
         }
 
-        let numerator = match numerator {
-            Some(num_str) => BigInt::from_str(num_str).map_err(IcuRatioError::BigIntParseError)?,
-            None => BigInt::zero(),
+        let significand = match significand {
+            Some(significand) => parse_fraction(significand)?,
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:308]: is the significant allowed to be a fraction when there's an exponent?

File: components/experimental/src/units/ratio.rs [hunk 5]
```diff
@@ -195,24 +208,184 @@ impl From<u32> for IcuRatio {
...
 
-        let denominator = match denominator {
-            Some(den_str) => BigInt::from_str(den_str).map_err(IcuRatioError::BigIntParseError)?,
-            None => BigInt::one(),
+        let exponent = match exponent {
+            Some(exponent) => {
+                let exponent = match exponent.parse::<i32>() {
+                    Ok(exponent) => exponent,
+                    Err(_) => return Err(IcuRatioError::ExponentPartIsNotInteger),
+                };
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:317]:
```suggestion
                let exponent = match exponent.parse::<i32>().map_err(|_| Err(IcuRatioError::ExponentPartIsNotInteger))?;
```

File: components/experimental/src/units/ratio.rs [hunk 6]
```diff
@@ -195,24 +208,184 @@ impl From<u32> for IcuRatio {
...
+                let ten = IcuRatio::ten();
+                ten.pow(exponent)
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:319]:
```suggestion
                IcuRatio::ten().pow(exponent)
```

File: components/experimental/src/units/ratio.rs (hunk 1 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:25]: do you mean that making an error depending on the functionality:

For example: `InvalidRatioString` or `ErrorInFromStr` ?

File: components/experimental/src/units/ratio.rs (hunk 4 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:308]: Yes, we can have `1/2E+5`

File: components/experimental/src/units/ratio.rs (hunk 1 above)
Comment (robertbastian) [reply components/experimental/src/units/ratio.rs:25]: Yes, `RatioFromStrError`

File: components/experimental/src/units/ratio.rs (hunk 4 above)
Comment (robertbastian) [reply components/experimental/src/units/ratio.rs:308]: please add this to the docs

File: components/experimental/src/units/ratio.rs (hunk 1 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:25]: done

File: components/experimental/src/units/ratio.rs (hunk 2 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs]: done

File: components/experimental/src/units/ratio.rs (hunk 4 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:297]: done

File: components/experimental/src/units/ratio.rs (hunk 5 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs]: done

```
let exponent = exponent
                    .parse::<i32>()
                    .map_err(|_| IcuRatioError::RatioFromStrError)?;
```

File: components/experimental/src/units/ratio.rs (hunk 6 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:308]: done

File: components/experimental/src/units/ratio.rs (hunk 1 above)
Comment (robertbastian) [reply components/experimental/src/units/ratio.rs:25]: I meant a different type `pub enum RatioFromStrError { DivisionByZero, MultipleSlashes, ... }`

Comment (younies) [reply components/experimental/src/units/ratio.rs:25]: OMG, that is a reverse change. 

let me try to do it

Comment (younies) [reply components/experimental/src/units/ratio.rs]: done

File: components/experimental/src/units/ratio.rs [hunk 7]
```diff
@@ -193,26 +211,194 @@ impl From<u32> for IcuRatio {
...
+    /// - Negative numbers with decimal notation: "-1.5" becomes "-3/2".
+    /// - Negative numbers with scientific notation: "-1.5E6" becomes "-1500000".
+    /// - Negative numbers with scientific notation with commas: "-1,500E6" becomes "-1500000000".
+    /// Parsing errors are returned for:
+    /// - Division by zero: "1/0".
+    /// - Multiple slashes: "1/2/3".
+    /// - Non-numeric characters in fractions: "1/2A".
+    /// - Multiple scientific notations: "1.5E6E6".
+    /// - Multiple decimal points: "1.5.6".
+    /// - Exponent part is not an integer: "1.5E6.5".
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:235]: suggestion: write these as doc tests to give more confidence that they are correct, and increase test coverage

File: components/experimental/src/units/ratio.rs [hunk 8]
```diff
@@ -193,26 +211,194 @@ impl From<u32> for IcuRatio {
...
+
+            Ok(integer_part + decimal_part)
+        }
+
+        let number_str = number_str.replace(',', ""); // Remove commas for parsing
+        if number_str.is_empty() {
+            return Ok(IcuRatio::zero());
+        }
+
+        let mut parts = number_str.split(|c| c == 'e' || c == 'E');
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:314]:
```suggestion
        let mut parts = number_str.split(['e', 'E']);
```

File: components/experimental/src/units/ratio.rs [hunk 9]
```diff
@@ -193,26 +211,194 @@ impl From<u32> for IcuRatio {
...
+            None => IcuRatio::one(),
         };
 
-        let denominator = match denominator {
-            Some(den_str) => BigInt::from_str(den_str).map_err(IcuRatioError::BigIntParseError)?,
-            None => BigInt::one(),
+        let exponent = match exponent {
+            Some(exponent) => {
+                let exponent = exponent
+                    .parse::<i32>()
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:329]: is `i32` always enough?

File: components/experimental/src/units/ratio.rs [hunk 10]
```diff
@@ -193,26 +211,194 @@ impl From<u32> for IcuRatio {
...
+                    );
+
+                    decimal_part / IcuRatio::ten().pow(decimal_power) // Divide by 10^decimal_power
+                }
+            };
+
+            Ok(integer_part + decimal_part)
+        }
+
+        let number_str = number_str.replace(',', ""); // Remove commas for parsing
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:309]: you can avoid the allocation in the general case:

```suggestion
        let number_str = if number_str.contains(',') {
            Cow::Owned(number_str.replace(',', ""))
        } else {
            Cow::Borrowed(number_str)
        };
```

File: components/experimental/src/units/ratio.rs [hunk 11]
```diff
@@ -193,26 +211,194 @@ impl From<u32> for IcuRatio {
...
+
+    /// Converts a string representation of a ratio into an `IcuRatio`.
+    /// Supported string formats include:
+    /// - Fractional notation: "1/2" becomes "1/2".
+    /// - Decimal notation: "1.5" becomes "3/2".
+    /// - Scientific notation: "1.5E6" becomes "1500000", "1.5E-6" becomes "0.0000015".
+    /// - Scientific notation with commas: "1,500E6" becomes "1500000000". Commas are disregarded.
+    /// - Integer notation: "1" becomes "1".
+    /// - Empty string: "" becomes "0".
+    /// - Fractional notation with exponent: "1/2E5" becomes "1/200000".
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:224]: issue: in the test this is `50000`

File: components/experimental/src/units/ratio.rs [hunk 12]
```diff
@@ -193,26 +211,194 @@ impl From<u32> for IcuRatio {
...
+                ),
+                (Err(ref actual_err), Err(ref expected_err)) => {
+                    assert_eq!(
+                        actual_err, expected_err,
+                        "Error types do not match for input: {}",
+                        input
+                    )
+                }
+                _ => panic!("Result types (Ok/Err) do not match for input: {}", input),
+            }
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:397]: does this not work?

```suggestion
            assert_eq!(
                actual, expected,
                "Values do not match for input: {}",
                input
            );
```

File: components/experimental/src/units/ratio.rs [hunk 13]
```diff
@@ -193,26 +211,194 @@ impl From<u32> for IcuRatio {
...
+        }
+
+        let actual = IcuRatio::from_str("1.5").unwrap();
+        let expected = IcuRatio::from_big_ints(3.into(), 2.into());
+        assert_eq!(actual, expected);
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:402]: why is this not in `test_cases`?

File: components/experimental/src/units/ratio.rs (hunk 9 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:334]: yes, if it is more, it will return an error

Comment (younies) [reply components/experimental/src/units/ratio.rs:334]: this is the exponent part

File: components/experimental/src/units/ratio.rs (hunk 13 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs]: good catch, I was just testing

File: components/experimental/src/units/ratio.rs (hunk 7 above)
Comment (robertbastian) [reply components/experimental/src/units/ratio.rs]: I would still like this

**Review by robertbastian — APPROVED** [review]

File: components/experimental/src/units/ratio.rs (hunk 7 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs]: Okay, I just saw it ,  I will do it in a follow-up PR.

**Pull request merged**